service/matching/physical_task_queue_manager.go

```diff
@@ -126,12 +132,109 @@ type (
 		clusterMeta          cluster.Metadata
 		taggedMetricsHandler metrics.Handler // namespace/taskqueue tagged metric scope
 		// pollerHistory stores poller which poll from this taskqueue in last few minutes
-		pollerHistory *pollerHistory
-		currentPolls  atomic.Int64
-		taskValidator taskValidator
+		pollerHistory              *pollerHistory
+		currentPolls               atomic.Int64
+		taskValidator              taskValidator
+		tasksAddedInIntervals      *taskTracker
+		tasksDispatchedInIntervals *taskTracker
 	}
 )
 
+// a circular array of a fixed size for tracking tasks
+type circularTaskBuffer struct {
+	buffer     []int
+	currentPos int
+}
+
+func newCircularTaskBuffer(size int) circularTaskBuffer {
+	return circularTaskBuffer{
+		buffer: make([]int, size),
+	}
+}
+
+func (cb *circularTaskBuffer) incrementTaskCount() {
+	cb.buffer[cb.currentPos]++
+}
+
+func (cb *circularTaskBuffer) advance() {
+	cb.currentPos = (cb.currentPos + 1) % len(cb.buffer)
+	cb.buffer[cb.currentPos] = 0 // Reset the task count for the new interval
+}
+
+// returns the total number of tasks in the buffer
+func (cb *circularTaskBuffer) totalTasks() int {
```

Comment:
I think it would be more efficient to keep a `totalTasks int` as a field of the `circularTaskBuffer` struct, and just increment / decrement it as part of `incrementTaskCount()` and `advance()`. It's common for a data structure to hold a "count" field so that calling `count()` (or in this case `totalTasks()`) can be O(1).

This works though, no need to change it, just sharing for your future knowledge